In [2]:
import pandas as pd

# Pré-processamento

### Ver informações para cada ano diferente

## 2015

In [3]:
ano_2015=pd.read_csv("2015.csv")
print(ano_2015.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        158 non-null    object 
 1   Region                         158 non-null    object 
 2   Happiness Rank                 158 non-null    int64  
 3   Happiness Score                158 non-null    float64
 4   Standard Error                 158 non-null    float64
 5   Economy (GDP per Capita)       158 non-null    float64
 6   Family                         158 non-null    float64
 7   Health (Life Expectancy)       158 non-null    float64
 8   Freedom                        158 non-null    float64
 9   Trust (Government Corruption)  158 non-null    float64
 10  Generosity                     158 non-null    float64
 11  Dystopia Residual              158 non-null    float64
dtypes: float64(9), int64(1), object(2)
memory usage: 1

## 2016

In [ ]:
ano_2016=pd.read_csv("2016.csv")
print(ano_2016.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        157 non-null    object 
 1   Region                         157 non-null    object 
 2   Happiness Rank                 157 non-null    int64  
 3   Happiness Score                157 non-null    float64
 4   Lower Confidence Interval      157 non-null    float64
 5   Upper Confidence Interval      157 non-null    float64
 6   Economy (GDP per Capita)       157 non-null    float64
 7   Family                         157 non-null    float64
 8   Health (Life Expectancy)       157 non-null    float64
 9   Freedom                        157 non-null    float64
 10  Trust (Government Corruption)  157 non-null    float64
 11  Generosity                     157 non-null    float64
 12  Dystopia Residual              157 non-null    flo

## 2017

In [ ]:
ano_2017=pd.read_csv("2017.csv")
print(ano_2017.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        155 non-null    object 
 1   Happiness.Rank                 155 non-null    int64  
 2   Happiness.Score                155 non-null    float64
 3   Whisker.high                   155 non-null    float64
 4   Whisker.low                    155 non-null    float64
 5   Economy..GDP.per.Capita.       155 non-null    float64
 6   Family                         155 non-null    float64
 7   Health..Life.Expectancy.       155 non-null    float64
 8   Freedom                        155 non-null    float64
 9   Generosity                     155 non-null    float64
 10  Trust..Government.Corruption.  155 non-null    float64
 11  Dystopia.Residual              155 non-null    float64
dtypes: float64(10), int64(1), object(1)
memory usage: 

## 2018

In [ ]:
ano_2018=pd.read_csv("2018.csv")
print(ano_2018.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     155 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB
None


## 2019

In [ ]:
ano_2019=pd.read_csv("2019.csv")
print(ano_2019.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB
None


## 2020

In [ ]:
ano_2020=pd.read_csv("2020.csv")
print(ano_2020.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 20 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country name                                153 non-null    object 
 1   Regional indicator                          153 non-null    object 
 2   Ladder score                                153 non-null    float64
 3   Standard error of ladder score              153 non-null    float64
 4   upperwhisker                                153 non-null    float64
 5   lowerwhisker                                153 non-null    float64
 6   Logged GDP per capita                       153 non-null    float64
 7   Social support                              153 non-null    float64
 8   Healthy life expectancy                     153 non-null    float64
 9   Freedom to make life choices                153 non-null    float64
 10  Generosity    

## 2021

In [ ]:
ano_2021=pd.read_csv("2021.csv")
print(ano_2021.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149 entries, 0 to 148
Data columns (total 20 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country name                                149 non-null    object 
 1   Regional indicator                          149 non-null    object 
 2   Ladder score                                149 non-null    float64
 3   Standard error of ladder score              149 non-null    float64
 4   upperwhisker                                149 non-null    float64
 5   lowerwhisker                                149 non-null    float64
 6   Logged GDP per capita                       149 non-null    float64
 7   Social support                              149 non-null    float64
 8   Healthy life expectancy                     149 non-null    float64
 9   Freedom to make life choices                149 non-null    float64
 10  Generosity    

## 2022

In [ ]:
ano_2022=pd.read_csv("2022.csv")
print(ano_2022.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147 entries, 0 to 146
Data columns (total 12 columns):
 #   Column                                      Non-Null Count  Dtype 
---  ------                                      --------------  ----- 
 0   RANK                                        147 non-null    int64 
 1   Country                                     147 non-null    object
 2   Happiness score                             146 non-null    object
 3   Whisker-high                                146 non-null    object
 4   Whisker-low                                 146 non-null    object
 5   Dystopia (1.83) + residual                  146 non-null    object
 6   Explained by: GDP per capita                146 non-null    object
 7   Explained by: Social support                146 non-null    object
 8   Explained by: Healthy life expectancy       146 non-null    object
 9   Explained by: Freedom to make life choices  146 non-null    object
 10  Explained by: Generosity  

### parte para mudar nomes de atributos e assim para ficar com as colunas todas iguais

- na coluna do ano nao temos d mudar nada
- no country- no d 2018/2019 existe country or region. no 2020/2021 é country name. nos outros é country
- region- 2017 nao tem region. 2018/2019 tem country or region. 2020/2021 tem regional indicator.  2022 nao tem region. o resto é region
- happines score- Happiness Score”, “Score”, “Ladder score”, “Happiness score” 
G- DP per capita- Economy..GDP.per.Capita, “Economy (GDP per Capita)”, “GDP per capita”, “Logged GDP per capita”, “Explained by: GDP per capita”
- social support- “Family”, “Social support”, “Explained by: Social support”
- healthy life expectancy- “Health (Life Expectancy)”, “Healthy life expectancy”, “Explained by: Healthy life expectancy”
- freedom- “Freedom”, “Freedom to make life choices”, “Explained by: Freedom to make life choices”
- corruption- “Trust (Government Corruption)”, “Perceptions of corruption”, “Explained by: Perceptions of corruption”
- gerenosity- “Generosity”, “Explained by: Generosity”
dystopia residual- “Dystopia Residual”, “Dystopia + residual”, “Dystopia (1.83) + residual”
- happiness rank- “Happiness Rank”, “Overall rank”, “RANK”

ha alguns anos que têm country e region tudo junto. oq fazemos aí vai ser com base nos outros anos que têm as colunas separadas qual é qual.
o chat dps disse isto: Depois, podes preencher regiões ausentes com base num dicionário país → região extraído de 2015/2016 ou 2020/2021.Dica: usar country-region mapping de anos com essa informação é totalmente válido e recomendado — as regiões raramente mudam



no gdp per capita temos dps d normalizar tudo 

se calhar nao utiliza estas colunas: 
Standard Error
Lower Confidence Interval
Upper Confidence Interval
Whisker.high / Whisker.low
Explained by
Ladder score in Dystopia

In [60]:
colunas_2015 = set(ano_2015.columns)
colunas_2016 = set(ano_2016.columns)

diferentes = colunas_2015.symmetric_difference(colunas_2016)

print("Colunas diferentes entre 2015 e 2016:")
print(diferentes)


Colunas diferentes entre 2015 e 2016:
{'Standard Error', 'Lower Confidence Interval', 'Upper Confidence Interval'}


In [74]:
colunas_2016 = set(ano_2016.columns)
colunas_2017 = set(ano_2017.columns)

diferentes = colunas_2016.symmetric_difference(colunas_2017)

print("Colunas diferentes entre 2016 e 2017:")
print(diferentes)

Colunas diferentes entre 2016 e 2017:
{'Whisker.low', 'Upper Confidence Interval', 'Region', 'Whisker.high', 'Lower Confidence Interval'}


In [75]:
colunas_2016 = set(ano_2016.columns)
colunas_2018 = set(ano_2018.columns)

diferentes = colunas_2016.symmetric_difference(colunas_2018)

print("Colunas diferentes entre 2016 e 2017:")
print(diferentes)

Colunas diferentes entre 2016 e 2017:
{'Happiness Rank', 'Social support', 'Score', 'Region', 'Perceptions of corruption', 'Happiness Score', 'Country', 'Economy (GDP per Capita)', 'GDP per capita', 'Upper Confidence Interval', 'Trust (Government Corruption)', 'Health (Life Expectancy)', 'Family', 'Healthy life expectancy', 'Freedom to make life choices', 'Freedom', 'Overall rank', 'Country or region', 'Lower Confidence Interval', 'Dystopia Residual'}


### Juntar os anos todos já com as colunas arranjadas e assim

### Valores em falta

### Dados duplicados

### Verificar tipo de dados

### Normalização

### Outliers

### Verificar consistência ao longo dos anos

### Resumo estatístico anual

### Histogramas e distribuições

### Balanceamento

### Matriz de correlação

-------------------------------------------------------------------------------

### **Pergunta geral**: 
Como evoluíram os fatores que influenciam a felicidade global entre 2015 e 2022, e que padrões regionais ou estruturais explicam essas mudanças?

### Primeira visualização
A felicidade mundial está a convergir ou a divergir ao longo do tempo?

### Segunda visualização
A pandemia de 2020 alterou os fatores que mais pesam na felicidade?

### Terceira visualização
A economia continua a ser o fator mais determinante da felicidade com o passar dos anos?

### Quarta visualização
Os países mais generosos tornaram-se mais ou menos felizes ao longo do tempo?

### Quinta visualização
Quais países mudaram mais o seu ranking de felicidade — e porquê?